File: crud-item-list/src/main/java/com/gurunars/crud_item_list/ActionDelete.java
package com.gurunars.crud_item_list;

import java.util.List;
import java.util.Set;

final class ActionDelete<ItemType> implements Action<ItemType> {

    @Override
    public boolean perform(List<ItemType> all, Set<ItemType> selectedItems) {
        all.removeAll(selectedItems);
        return true;
    }

    @Override
    public boolean canPerform(List<ItemType> all, Set<ItemType> selectedItems) {
        return selectedItems.size() > 0;
    }

}

File: crud-item-list/src/main/kotlin/com/gurunars/crud_item_list/ActionDelete.kt
package com.gurunars.crud_item_list

import com.gurunars.box.IRoBox
import com.gurunars.box.oneWayBranch
import com.gurunars.item_list.Item

internal class ActionDelete<ItemType : Item> : Action<ItemType> {

    override fun canPerform(
        all: IRoBox<List<ItemType>>,
        selectedItems: IRoBox<Set<ItemType>>
    ): IRoBox<Boolean> = selectedItems.oneWayBranch { isNotEmpty() }

    override fun perform(
        all: List<ItemType>,
        selectedItems: Set<ItemType>,
        consumer: ItemSetChange<ItemType>
    ) = consumer(
        all.filter { item -> selectedItems.indexOfFirst { it.id == item.id } == -1 },
        setOf()
    )
}
